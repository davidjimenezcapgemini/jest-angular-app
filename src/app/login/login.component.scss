.form-input {
    display: flex;
    flex-direction: column;
    margin-bottom: 20px;
}

.label {
    font-weight: 500;
    line-height: 3;
    font-size: 16px;
}

.general-input {
    font-size: 14px;
    padding: 12px 5px;
    outline: none;
    background-color: #1976d2;
    color: white;
    border: 0;
    border-radius: 4px;
}

.submit-button {
    font-size: 14px;
    padding: 12px 5px;
    outline: none;
    background-color: #1976d2;
    color: white;
    border: 0;
    border-radius: 4px;
    text-align: center;
    cursor: pointer;
    margin-top: 40px;
    font-weight: 600;

    &:disabled {
        opacity: .6;
    }
}

.error-input {
    color: rgb(221, 0, 49);
    font-weight: 500;
    margin-top: 4px;
}



