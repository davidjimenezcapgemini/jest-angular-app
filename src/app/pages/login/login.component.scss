.form-input {
    display: flex;
    flex-direction: column;
    margin-bottom: 20px;
}

.label {
    font-weight: 500;
    line-height: 3;
    font-size: 16px;
}

.general-input {
    font-size: 14px;
    padding: 12px 5px;
    outline: none;
    background-color: #1976d2;
    color: white;
    border: 0;
    border-radius: 4px;
}

.submit-button {
    font-size: 14px;
    padding: 12px 5px;
    outline: none;
    background-color: #1976d2;
    color: white;
    border: 0;
    border-radius: 4px;
    text-align: center;
    cursor: pointer;
    margin-top: 40px;
    font-weight: 600;

    &:disabled {
        opacity: .6;
    }
}

.error-input {
    color: rgb(221, 0, 49);
    font-weight: 500;
    margin-top: 4px;
}

.spinner {
    width: 100vw;
    height: 100vh;
    position: fixed;
    top: 0;
    left: 0;
    background-color: cadetblue;
    opacity: .6;
}



