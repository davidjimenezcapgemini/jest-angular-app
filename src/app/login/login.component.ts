import { LoginService } from './../services/login.service';
import { Component, OnInit } from '@angular/core';
import { FormControl, FormGroup, Validators } from '@angular/forms';
import { Router } from '@angular/router';

@Component({
  selector: 'app-login',
  templateUrl: './login.component.html',
  styleUrls: ['./login.component.scss']
})
export class LoginComponent implements OnInit {

  loginForm: FormGroup;
  spinner: boolean;

  constructor(private loginService: LoginService, private router: Router) {
    this.loginForm = new FormGroup({
      username: new FormControl('', Validators.required),
      password: new FormControl('', Validators.required)
    });
    this.spinner = false;
  }

  ngOnInit(): void {
  }

  submit(): void {
    if (this.loginForm.valid) {
      this.spinner = true;
      const user = this.loginForm.get('username')?.value;
      const pass = this.loginForm.get('password')?.value;
      this.loginService.login(user, pass).subscribe((res: any) => {
        const jwt = res && res.data ? res.data.jwt : null; // Solution: Remove to do exercise
        this.loginService.setLogged(jwt);
        this.spinner = false;
        this.router.navigate(['dashboard']);
      });
    }
  }

}
