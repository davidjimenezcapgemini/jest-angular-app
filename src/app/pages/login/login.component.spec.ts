import { LoginService } from '../../services/login.service';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';

import { LoginComponent } from './login.component';
import {} from 'jest-image-snapshot';
import { Observable } from 'rxjs';
import { Router } from '@angular/router';
import { RouterTestingModule } from '@angular/router/testing';

describe('LoginComponent snapshots test', () => {
  let component: LoginComponent;
  let fixture: ComponentFixture<LoginComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ LoginComponent ],
      imports: [ReactiveFormsModule, RouterTestingModule],
      providers: [LoginService]
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(LoginComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  test('should matches the current snapshot with component', () => {
    expect(fixture).toMatchSnapshot();
  });
});

class RouterStub {
    url = '';
    // tslint:disable-next-line:typedef
    navigate(commands: any[], extras?: any) { }
  }

describe('LoginComponent functions', () => {
    let component: LoginComponent;
    let fixture: ComponentFixture<LoginComponent>;
    let loginService: LoginService;

    beforeEach(async () => {
        await TestBed.configureTestingModule({
        declarations: [ LoginComponent ],
        imports: [ReactiveFormsModule, RouterTestingModule],
        providers: [{ provide: Router, useClass: RouterStub }]
        })
        .compileComponents();
    });

    beforeEach(() => {
        fixture = TestBed.createComponent(LoginComponent);
        component = fixture.componentInstance;
        loginService = TestBed.inject(LoginService);
        fixture.detectChanges();
    });

    afterEach(() => {
        jest.clearAllMocks();
        sessionStorage.clear();
    });

    test('Should does nothing when form is invalid', () => {
        const spy = jest.spyOn(loginService, 'login');
        component.submit();
        expect(spy).not.toHaveBeenCalled();
        expect(component.spinner).toEqual(false);
    });

    test('Should does login when form is valid', (done) => {
        component.loginForm.setValue({username: 'test', password: 'passtest'});
        component.submit();
        expect(component.spinner).toEqual(true);
        setTimeout(() => {
        expect(component.spinner).toEqual(false);
        const jwt = sessionStorage.getItem('jwt');
        expect(jwt).not.toBeNull();
        done();
        }, 3000);
    });

    test('Should does not save jwt when service does not return it', (done) => {
        component.loginForm.setValue({username: 'test', password: 'passtest'});
        const $obs = new Observable<object>(observer => {
        setTimeout(() => {
            observer.next(undefined);
        }, 100);
        });
        const spy = jest.spyOn(loginService, 'login').mockReturnValue($obs);
        component.submit();
        setTimeout(() => {
        const jwt = sessionStorage.getItem('jwt');
        expect(jwt).toStrictEqual('null');
        done();
        }, 500);
    });

});
