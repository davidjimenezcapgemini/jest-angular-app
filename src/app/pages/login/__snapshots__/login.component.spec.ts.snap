// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginComponent snapshots test should matches the current snapshot with component 1`] = `
<app-login
  loginForm={[Function FormGroup]}
  loginService={[Function LoginService]}
  router={[Function Router]}
  spinner="false"
>
  <h1>
    Login View
  </h1><form
    class="ng-untouched ng-pristine ng-invalid"
    novalidate=""
  >
    <div
      class="form-input"
    >
      <label
        class="label"
      >
        Username
      </label>
      <input
        class="general-input ng-untouched ng-pristine ng-invalid"
        formcontrolname="username"
        type="text"
      />
      
      <span
        class="error-input"
      >
        Required
      </span>
    </div>
    <div
      class="form-input"
    >
      <label
        class="label"
      >
        Password
      </label>
      <input
        class="general-input ng-untouched ng-pristine ng-invalid"
        formcontrolname="password"
        type="password"
      />
      
      <span
        class="error-input"
      >
        Required
      </span>
    </div>
    <div
      class="form-input"
    >
      <button
        class="submit-button"
        disabled=""
      >
        Enter
      </button>
    </div>
  </form>
</app-login>
`;
